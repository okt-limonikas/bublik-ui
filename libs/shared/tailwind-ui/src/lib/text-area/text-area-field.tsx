/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { Control, FieldValues, Path, useController } from 'react-hook-form';

import { TextArea, TextAreaProps } from './text-area';

export interface TextAreaFieldProps<T extends FieldValues>
	extends TextAreaProps {
	name: Path<T>;
	control: Control<T, unknown>;
}

export const TextAreaField = <T extends FieldValues>(
	props: TextAreaFieldProps<T>
) => {
	const { field, fieldState } = useController({
		name: props.name,
		control: props.control
	});

	return <TextArea error={fieldState.error?.message} {...field} {...props} />;
};
