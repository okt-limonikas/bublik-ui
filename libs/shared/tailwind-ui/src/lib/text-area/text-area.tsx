/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import React, { ComponentPropsWithRef, forwardRef } from 'react';

import { cn, cva, VariantProps } from '../utils';

const textAreaStyles = cva({
	base: 'block w-full rounded-md sm:text-sm focus:ring-transparent transition-all resize-none active:shadow-none',
	variants: {
		variant: {
			primary:
				'border-border-primary focus:shadow-text-field focus:border-primary hover:border-primary',
			error:
				'border-bg-error focus:shadow-text-field-error focus:border-bg-error hover:border-bg-error'
		}
	}
});

export type TextAreaProps = ComponentPropsWithRef<'textarea'> &
	VariantProps<typeof textAreaStyles> & {
		label: string;
		error?: string;
	};

export const TextArea = forwardRef<HTMLTextAreaElement, TextAreaProps>(
	({ error, name, label, variant = 'primary', className, ...props }, ref) => {
		return (
			<div className="flex flex-col gap-2">
				<label
					htmlFor={name}
					className="block mb-1 text-sm font-medium text-text-primary"
				>
					{label}
				</label>
				<textarea
					{...props}
					name={name}
					id={name}
					className={cn(
						textAreaStyles({ variant: error ? 'error' : variant }),
						className
					)}
					ref={ref}
				/>
				{error && (
					<p className="text-[0.75rem] font-normal text-bg-error">{error}</p>
				)}
			</div>
		);
	}
);
