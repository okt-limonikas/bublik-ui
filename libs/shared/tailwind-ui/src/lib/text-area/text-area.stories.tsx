/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import type { Meta } from '@storybook/react';
import { withBackground } from '../storybook-bg';

import { TextArea } from './text-area';

const Story: Meta<typeof TextArea> = {
	component: TextArea,
	title: 'components/Text Area',
	decorators: [withBackground]
};
export default Story;

export const Primary = {
	args: {
		label: 'Label',
		name: 'name',
		placeholder: 'Placeholder'
	}
};
