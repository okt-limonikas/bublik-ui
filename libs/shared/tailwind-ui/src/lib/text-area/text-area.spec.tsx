/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { render } from '@testing-library/react';

import { TextArea } from './text-area';

describe('TextArea', () => {
	it('should render successfully', () => {
		const { asFragment } = render(<TextArea label={'Text area'} />);

		expect(asFragment()).toBeTruthy();
	});

	it('should render default variant', () => {
		const { asFragment } = render(<TextArea label={'Text area'} />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should render primary variant', () => {
		const { asFragment } = render(
			<TextArea label={'Text area'} variant="primary" />
		);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should render with error', () => {
		const { asFragment } = render(
			<TextArea label={'Text area'} variant="primary" error={'Error message'} />
		);

		expect(asFragment()).toMatchSnapshot();
	});
});
