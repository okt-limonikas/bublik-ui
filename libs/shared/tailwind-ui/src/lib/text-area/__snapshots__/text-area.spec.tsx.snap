// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextArea > should render default variant 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-2"
  >
    <label
      class="block mb-1 text-sm font-medium text-text-primary"
    >
      Text area
    </label>
    <textarea
      class="block w-full rounded-md sm:text-sm focus:ring-transparent transition-all resize-none active:shadow-none border-border-primary focus:shadow-text-field focus:border-primary hover:border-primary"
    />
  </div>
</DocumentFragment>
`;

exports[`TextArea > should render primary variant 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-2"
  >
    <label
      class="block mb-1 text-sm font-medium text-text-primary"
    >
      Text area
    </label>
    <textarea
      class="block w-full rounded-md sm:text-sm focus:ring-transparent transition-all resize-none active:shadow-none border-border-primary focus:shadow-text-field focus:border-primary hover:border-primary"
    />
  </div>
</DocumentFragment>
`;

exports[`TextArea > should render with error 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-2"
  >
    <label
      class="block mb-1 text-sm font-medium text-text-primary"
    >
      Text area
    </label>
    <textarea
      class="block w-full rounded-md sm:text-sm focus:ring-transparent transition-all resize-none active:shadow-none border-bg-error focus:shadow-text-field-error focus:border-bg-error hover:border-bg-error"
    />
    <p
      class="text-[0.75rem] font-normal text-bg-error"
    >
      Error message
    </p>
  </div>
</DocumentFragment>
`;
